node/network/src/proxy/benchmark.rs

```diff
@@ -0,0 +1,48 @@
+use futures::Stream;
+use rand::Rng;
+
+use crate::protocol::Package;
+use crate::proxy::ProxyHandler;
+use std::time::Instant;
+use std::sync::RwLock;
+use std::sync::Arc;
+use std::cmp::min;
+
+
+pub struct BenchmarkHandler {
+    instants: Arc<RwLock<Vec<Instant>>>
+}
+
+impl BenchmarkHandler {
+    pub fn new() -> Self {
+        Self {
+            instants: Arc::new(RwLock::new(vec![]))
+        }
+    }
+}
+
+/// Messages will be dropped with probability `dropout_rate `
+impl ProxyHandler for BenchmarkHandler {
+    fn pipe_stream(&self, stream: Box<Stream<Item=Package, Error=()> + Send + Sync>) ->
+    Box<Stream<Item=Package, Error=()> + Send + Sync>
+    {
+        Box::new(stream.map(|package| {
+            let now = Instant::now();
+
+            self.instants.write().expect("The lock was poisoned.").push(now);
```

Comment:
@nearmax I can't access `self` in `pipe_stream` without requiring handling lifetimes explicitly. This handler [is disabled](https://github.com/nearprotocol/nearcore/pull/735/files#diff-29fd0ddb6da41de684f5e203278599c9R17) right now.

Reply:
The closure that you are passing to `stream.map` is capturing the variables in the scope. The rule is that it first tries to capture them by reference. There is only one variable in the body of `pipe_stream` that ends up being used in `stream.map`. It is `self`, which is the reference with some lifetime. However, we then wrap this closure in `Box` and pass outside which means it is undetermined when this closure is going to be used, so it demands `static` lifetime from `self`.

The solution is:

Create:
```
let instants = self.instants.clone();
Box::new(stream.map(move |package| {
instants.write().....
```

Reply:
thanks!